sumtype not being auto converted back to original type inside match statement and C error

### Describe the bug

The sumtype is not automatically being converted back into its original type while inside match statement which is resulting in a C error.

### Expected Behavior

Expected code to compile and run normally.

### Current Behavior

```
==================
C:/Users/imado/AppData/Local/Temp/v_0/main.14614145064292582856.tmp.c:6840: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/imado/AppData/Local/Temp/v_0/main.14614145064292582856.tmp.c:12606: error: cannot convert 'struct main__RuneAliasOrOtherType' to 'unsigned int'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

```v
struct OtherType {
    foo string
    bar string
}
type RuneAlias = rune
type RuneAliasOrOtherType = OtherType | RuneAlias

const whitespace = [rune(0x0009), 0x000a, 0x000c, 0x000d, 0x0020]

fn main() {
    mut x := RuneAliasOrOtherType(RuneAlias(rune(`!`)))
    match x {
        RuneAlias {
            if x in whitespace {
                // doing `if x as RuneAlias in whitepsace` here
                // works but it should be doing that automatically
                // since I'm inside the RuneAlias match condition.
            }
        }
        else {
            // do something else
        }
    }
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.0 76b4c92.c862394

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.0 76b4c92.c862394
OS: windows, Microsoft Windows 11 Pro v22621 64-bit
Processor: 16 cpus, 64bit, little endian,

getwd: C:\Users\imado\Documents\cyberian_tiger
vexe: C:\Users\imado\v\v.exe
vexe mtime: 2023-08-17 20:42:38

vroot: OK, value: C:\Users\imado\v
VMODULES: OK, value: C:\Users\imado\.vmodules
VTMP: OK, value: C:\Users\imado\AppData\Local\Temp\v_0

Git version: git version 2.33.1.windows.1
Git vroot status: weekly.2023.32-71-gc8623943
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -190,6 +190,12 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 							for i, typ in node.right.expr_types {
 								c.ensure_type_exists(typ, node.right.exprs[i].pos())
 							}
+						} else {
+							elem_type := right_final_sym.array_info().elem_type
+							c.check_expected(left_type, elem_type) or {
+								c.error('left operand to `${node.op}` does not match the array element type: ${err.msg()}',
+									left_right_pos)
+							}
 						}
 					}
 				}

--- a/vlib/v/checker/tests/infix_sumtype_in_array_err.out
+++ b/vlib/v/checker/tests/infix_sumtype_in_array_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/infix_sumtype_in_array_err.vv:15:7: error: left operand to `in` does not match the array element type: expected `rune`, not `RuneAliasOrOtherType`
+   13 |     match x {
+   14 |         RuneAlias {
+   15 |             if x in whitespace {
+      |                ~~~~~~~~~~~~~~~
+   16 |                 // doing `if x as RuneAlias in whitepsace` here
+   17 |                 // works but it should be doing that automatically

--- a/vlib/v/checker/tests/infix_sumtype_in_array_err.vv
+++ b/vlib/v/checker/tests/infix_sumtype_in_array_err.vv
@@ -1,0 +1,25 @@
+struct OtherType {
+	foo string
+	bar string
+}
+
+type RuneAlias = rune
+type RuneAliasOrOtherType = OtherType | RuneAlias
+
+const whitespace = [rune(0x0009), 0x000a, 0x000c, 0x000d, 0x0020]
+
+fn main() {
+	mut x := RuneAliasOrOtherType(RuneAlias(rune(`!`)))
+	match x {
+		RuneAlias {
+			if x in whitespace {
+				// doing `if x as RuneAlias in whitepsace` here
+				// works but it should be doing that automatically
+				// since I'm inside the RuneAlias match condition.
+			}
+		}
+		else {
+			// do something else
+		}
+	}
+}

